compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -860,6 +860,29 @@ impl<'tcx> PolyTraitRef<'tcx> {
     pub fn def_id(&self) -> DefId {
         self.skip_binder().def_id
     }
+
+    /// Fall back to `!const` if the `other` is `~const`
+    pub fn pick_concrete_effect(self, other: Self, constness: ty::BoundConstness) -> Option<Self> {
+        if self.def_id() != other.def_id() {
+            return None;
+        }
+        if self.self_ty() != other.self_ty() {
+            return None;
+        }
+        // FIXME(effects): somewhat fishy, should we do a step by step walk of the
+        // two substs and compare at each site?
+        match (self.has_effect_param(), other.has_effect_param(), constness) {
```

Comment:
it definitely seems weird to completely ignore every other argument apart from `Self`